[1.21.3] `RegisterRenderStateModifiersEvent` for appending custom data to render state objects

dhyces opened the pull request:
Closes #1638 

This PR targets extensibility of `EntityRenderState`, `MapRenderState`, and `MapDecorationRenderState` due to how inextensible the new render state system is. The system added on top allows for adding client-sided elements to render states. To accomplish this, a new event is added `RegisterRenderStateModifiersEvent`. The event supplies several methods for registering modifiers for the various types of render state objects. Should be extended with another method `registerItemModifier` in 1.21.4 for `ItemStackRenderState`s. Users would then add to the render state by calling `IRenderStateExtension#setRenderData` with a statically initialized `ContextKey` and the data they wish to store. Vanilla data may also be modified (at the user's discretion). For entity render states, every modifier that applies to the given renderer class is run, including modifiers that target superclasses, and are cached to be as efficient as possible.

dhyces committed:
Cherry-pick changes from example 2 branch

dhyces committed:
Draft for event

dhyces committed:
Example 1

dhyces committed:
Example 2

Technici4n commented:
I think that having:
- The attachment holder (which is weird if it's mutable I think).
- The ad-hoc key system to add extensions.
- The extension/merger system.
is 3 ways of achieving essentially the same thing. That's too much. I think there should only be one way.

Also, are there thread-safety or similar issues that would force us to make the data immutable? Or is it fine to directly reference the entity or something that it contains?

dhyces commented:
> I think that having:
> - The attachment holder (which is weird if it's mutable I think).
> - The ad-hoc key system to add extensions.
> - The extension/merger system.
> is 3 ways of achieving essentially the same thing. That's too much. I think there should only be one way.
> 
> Also, are there thread-safety or similar issues that would force us to make the data immutable? Or is it fine to directly reference the entity or something that it contains?

I mentioned in the description that the key system and extension system were two different examples for how it could work, only one would be implemented. I somewhat agree that the attachments would be enough (and yes, while it's mutable, it's a copy and wouldn't affect the client entity), though there may be some aspects that simply would be better off not using attachments and should use something else to extend the RenderState. Tslat mentioned in neocord that generic extension outside of attachments would be necessary.

Technici4n commented:
An idea might be to use a context map 🤔

dhyces committed:
Apply Tslat's feedback

1. Moved callsite to `createRenderState`. This differs from feedback, though I believe is a better location, seeing as the method is final and will apply modifiers if users call the method.
2. Some classes have been renamed to better reflect behavior
3. Use ContextKey instead of custom class
4. Retain each modifier separately so that subclasses may be considered
5. Use lazily evaluated map for storing all modifiers that apply to a given renderer

dhyces committed:
Rename to `RegisterRenderStateModifiersEvent`

XFactHD commented on patches/net/minecraft/client/renderer/entity/state/EntityRenderState.java.patch:
```diff
@@ -1,10 +1,39 @@
      @Nullable
      public EntityRenderState.LeashState leashState;
 +    public float partialTick;
++    private java.util.Map<net.minecraft.util.context.ContextKey<?>, Object> extensions = java.util.Map.of(); // NeoForge: allow generic extension of render states
```
There should be a way to clear this map since the render state is shared and people will only write to it when they need to.

dhyces replied:
I can't think of any cases where it would be worth clearing tbh. Do you have an example of when it would be cleared?

XFactHD replied:
It would be cleared every time the render state extraction happens, right before the modifiers are called. This would prevent state leaks between entities when the extension is conditional as I would expect people to just do something like this:
```java
Thing thing = computeData(entity);
if (thing != null)
    state.setRenderData(KEY, thing);
```
instead of always writing the result to the extension.

dhyces replied:
ah, rather than just passing null and removing the object. Sure, I suppose for safety reasons it makes sense to do that.

XFactHD commented on patches/net/minecraft/client/renderer/entity/state/EntityRenderState.java.patch:
```diff
@@ -1,10 +1,39 @@
++    @Override
++    public <T> void setRenderData(net.minecraft.util.context.ContextKey<T> key, @Nullable T data) {
++        if (extensions == java.util.Map.<net.minecraft.util.context.ContextKey<?>, Object>of()) {
++            extensions = new it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap<>();
```
I'm not sure this memory optimization is worth it considering there should only be one render state per renderer.

dhyces replied:
Sure, my thought process was that most people would likely be targeting a specific renderer or subset of renderers and there likely would not be many that put this system to use, thus it would be wasteful to instantiate a new map for all of them

XFactHD commented on src/main/java/net/neoforged/neoforge/client/RenderStateExtensions.java:
```diff
@@ -0,0 +1,48 @@
+        return (Collection<EntityRenderStateModifier<E, S>>) (Object) ENTITY_CACHE.computeIfAbsent((Class<? extends EntityRenderer<E, S>>) renderer.getClass(), aClass -> {
+            var list = new ObjectArrayList<EntityRenderStateModifier<?, ?>>();
+            for (var entry : ENTITY_EXTENSIONS.entrySet()) {
+                if (aClass.isInstance(entry.getClass())) {
```
This should be using `entry.getKey()` instead of `entry.getClass()` since the latter is just the relevant subclass of `Map.Entry`.

dhyces replied:
Whoops, thanks for that catch

dhyces committed:
Spotless

dhyces committed:
Use appropriate class from the entry

dhyces committed:
Reset render data before computing modifiers

XFactHD commented on patches/net/minecraft/client/renderer/entity/state/EntityRenderState.java.patch:
```diff
@@ -1,10 +1,44 @@
++
++    @org.jetbrains.annotations.ApiStatus.Internal
++    public void resetRenderData() {
++        extensions.clear();
```
This also needs to check for the map not being `Map.of()`, otherwise it'll throw an `UnsupportedOperationException`

dhyces committed:
Update EntityRenderState.java.patch

dhyces committed:
Remove lazy map init

dhyces committed:
Merge branch '1.21.x' into feat/render-state-event

XFactHD reviewed (commented):
The overall approach seems sane to me. One thing I'm thinking about though is whether it would make sense to make these extensions a bit more generic from the start by using two base classes for them which the vanilla ones extend from instead of patching the entire machinery into each of the respective vanilla "base" render state class (in this case `EntityRenderState`). What I'm imagining is a `BaseRenderState` or `AbstractRenderState` that holds the generic extension part and a `AttachmentAwareRenderState` which extends the former and implements `IAttachmentHolder`. The attachment holder implementation would be immutable which may also allow using an unmodifiable view of the entity's attachments map instead of copying it. The vanilla `EntityRenderState` (and the respective class for BEs if BERs get the same treatment) would then extend `AttachmentAwareRenderState` while `MapRenderState`, `MapDecorationRenderState` and `ItemStackRenderState` would extend `BaseRenderState`. The state modifier registration and retrieval would of course still have to be per "base type" (i.e. entity, map, map decoration, BE, itemstack).

dhyces commented:
> The overall approach seems sane to me. One thing I'm thinking about though is whether it would make sense to make these extensions a bit more generic from the start by using two base classes for them which the vanilla ones extend from instead of patching the entire machinery into each of the respective vanilla "base" render state class (in this case `EntityRenderState`). What I'm imagining is a `BaseRenderState` or `AbstractRenderState` that holds the generic extension part and a `AttachmentAwareRenderState` which extends the former and implements `IAttachmentHolder`. The attachment holder implementation would be immutable which may also allow using an unmodifiable view of the entity's attachments map instead of copying it. The vanilla `EntityRenderState` (and the respective class for BEs if BERs get the same treatment) would then extend `AttachmentAwareRenderState` while `MapRenderState`, `MapDecorationRenderState` and `ItemStackRenderState` would extend `BaseRenderState`. The state modifier registration and retrieval would of course still have to be per "base type" (i.e. entity, map, map decoration, BE, itemstack).

I like this idea a lot. I haven't yet looked into the `Map...` render states, so I'm not sure how that could be useful. Could you provide an example? I also have a concern with the attachment stuff, since we provide the arbitrary render data already, should we just tell users that if they want to access some data from their synced attachments, they can use a modifier to copy it over themselves? That way everything's not copied every time regardless of whether it's used or not.

XFactHD commented:
> I haven't yet looked into the `Map...` render states, so I'm not sure how that could be useful. Could you provide an example?

I don't have a specific example in mind but the uses of `MapItemSavedData` accessors are heavily patched to allow mods to provide custom data for special maps and I'm aware of at least one mod making use of this. While there is currently no way to modify the rendering of the map itself outside of tricks with the color values stored in the map data, the rendering of custom map decorations is customizable by registering an `IMapDecorationRenderer`. The custom renderer receives the `MapRenderState.MapDecorationRenderState` and can render whatever it wants, which may include dynamic things based on state extracted from the map data (which is currently not possible because the `MapItemSavedData` is no longer passed through to the decoration renderer).

> I also have a concern with the attachment stuff, since we provide the arbitrary render data already, should we just tell users that if they want to access some data from their synced attachments, they can use a modifier to copy it over themselves? That way everything's not copied every time regardless of whether it's used or not.

That's a good point, I would be in favor of that. It would also save us from potentially having to introduce a `DataComponent`-aware version for items once `ItemStackRenderState` is a thing (regardless of whether it would be another "base" class or directly patched into the `ItemStackRenderState`).

dhyces committed:
Merge branch '1.21.x' into feat/render-state-event

dhyces committed:
Update with proposed changes and add jdocs

dhyces committed:
Spotless

XFactHD commented on src/main/java/net/neoforged/neoforge/client/extensions/IRenderStateExtension.java:
```diff
@@ -0,0 +1,65 @@
+ * Extension class for render state objects. Implemented by {@link net.neoforged.neoforge.client.BaseRenderState} for
+ * simple class extension.
+ */
+public interface IRenderStateExtension {
```
Is there any value in keeping this interface around instead of folding it into `BaseRenderState`?

dhyces replied:
I figured it would be good to keep around for the future, in case there are ever any render state objects that extend some superclass

XFactHD replied:
Fair enough

dhyces committed:
Remove rather unnecessary interface indirection

dhyces committed:
Missed patch generation

dhyces committed:
Fix entity render states

dhyces committed:
Fail instead of throw

dhyces committed:
Add another fail condition

dhyces committed:
Add map render state modifiers

dhyces committed:
Spotless

dhyces committed:
Inline

Forgot I had removed the rest of the method

XFactHD commented on patches/net/minecraft/client/renderer/MapRenderer.java.patch:
```diff
@@ -11,7 +11,17 @@
+         p_364922_.texture = this.mapTextureManager.prepareMapTexture(p_361383_, p_363500_);
+         p_364922_.decorations.clear();
+-
++        net.neoforged.neoforge.client.ClientHooks.onUpdateMapRenderState(p_363500_, p_364922_);
```
Very minor nitpick: this can be a pure addition by keeping the empty line.

dhyces replied:
Good point, thanks

XFactHD commented on src/main/java/net/neoforged/neoforge/client/BaseRenderState.java:
```diff
@@ -0,0 +1,40 @@
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <T> T getRenderData(ContextKey<T> key) {
+        return (T) extensions.get(key);
+    }
+
+    public <T> void setRenderData(ContextKey<T> key, @Nullable T data) {
```
Minor nitpick: please add the missing override annotations.

dhyces replied:
Whoops, I had initially removed the extension interface, forgot to add Override on them since it's usually handled automatically

XFactHD commented on patches/net/minecraft/client/renderer/MapRenderer.java.patch:
```diff
@@ -11,7 +11,17 @@
++        net.neoforged.neoforge.client.ClientHooks.onUpdateMapRenderState(p_363500_, p_364922_);
+         for (MapDecoration mapdecoration : p_363500_.getDecorations()) {
+-            p_364922_.decorations.add(this.extractDecorationRenderState(mapdecoration));
++            p_364922_.decorations.add(net.neoforged.neoforge.client.ClientHooks.onUpdateMapDecorationRenderState(mapdecoration.type(), p_363500_, p_364922_, this.extractDecorationRenderState(mapdecoration)));
```
I think it would be useful to pass the entire `MapDecoration` into the modifier in case the decoration's position, rotation or name is relevant in same way.

dhyces replied:
`MapDecorationRenderState` contains that information already

XFactHD replied:
Right, I completely forgot about that.

XFactHD reviewed (changes requested):
I think from an organizational standpoint it would be nice to move `BaseRenderState`, `RenderStateExtensions` and `RegisterRenderStateModifiersEvent` into a `neoforge/client/renderstate` subpackage, move the `MapDecorationRenderStateModifier` to a separate class, move the update methods from `ClientHooks` into `RenderStateExtensions` and make the `RenderStateExtensions.registerX()` methods package-private. The `client` package is starting to get a bit bloated with a bunch of seemingly random classes like the `common` package and I'd prefer if we got away from just throwing everything into these two "catch-all" packages.

dhyces commented:
> I think from an organizational standpoint it would be nice to move... `RegisterRenderStateModifiersEvent` into a `neoforge/client/renderstate` subpackage...

hmmm, are there any other events that are outside of the event package? Don't really like breaking that expected convention

dhyces committed:
Move some classes to specific package

dhyces committed:
Split out MapDecorationRenderStateModifier

XFactHD commented:
Yes, there are already a couple events outside the common and client `events` packages, a couple examples being `RegisterCapabilitiesEvent`, `RegisterClientExtensionsEvent` and all registry-related events

dhyces committed:
Add override annotations

dhyces committed:
Move update methods into RenderStateExtensions

dhyces committed:
Move event into subpackage and reduce method vis

dhyces committed:
Apply formatting

dhyces committed:
Generate patches

dhyces committed:
Merge branch '1.21.x' into feat/render-state-event

XFactHD commented on src/main/java/net/neoforged/neoforge/client/renderstate/BaseRenderState.java:
```diff
@@ -0,0 +1,42 @@
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Extension class for RenderState objects (ie {@link net.minecraft.client.renderer.entity.state.EntityRenderState}).
```
Prefer imports over FQNs for docs in Neo classes

XFactHD commented on src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    /**
+     * Registers a {@link BiConsumer} for use in-game when updating {@link net.minecraft.client.renderer.state.MapRenderState}s. Can
+     * add custom data to the map using {@link net.neoforged.neoforge.client.extensions.IRenderStateExtension#setRenderData(ContextKey, Object)}.
+     * Modifiers are run after the texture has been set and before decorations have been added.
+     *
+     * @param modifier The function for modifying the {@link net.minecraft.client.renderer.state.MapRenderState} and adding custom render data.
```
Prefer imports over FQNs for docs in Neo classes

XFactHD commented on src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    /**
+     * Registers a {@link BiConsumer} for use in-game when updating {@link net.minecraft.client.renderer.state.MapRenderState.MapDecorationRenderState}s. Can
+     * add custom data to the map using {@link net.neoforged.neoforge.client.extensions.IRenderStateExtension#setRenderData(ContextKey, Object)}.
+     * Modifiers are run after vanilla map decoration data has been set.
+     *
+     * @param mapDecorationTypeKey Key for the registered {@link MapDecorationType}
+     * @param modifier             The function for modifying the {@link net.minecraft.client.renderer.state.MapRenderState.MapDecorationRenderState} and adding custom render data.
```
Prefer imports over FQNs for docs in Neo classes

XFactHD commented on src/main/java/net/neoforged/neoforge/client/renderstate/MapDecorationRenderStateModifier.java:
```diff
@@ -0,0 +1,13 @@
+import net.minecraft.client.renderer.state.MapRenderState;
+import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
+
+public interface MapDecorationRenderStateModifier {
```
Please add a brief doc comment

XFactHD reviewed (changes requested):
A couple more points about docs, then it's good to go from my side.

GizmoTheMoonPig reviewed (approved):
System looks good to me and seems to cover all the things I still need to fix for porting. Thanks for all your work on this, its greatly appreciated

dhyces committed:
Docs changes

dhyces committed:
Simplify RenderStateExtensions

XFactHD commented on src/main/java/net/neoforged/neoforge/client/renderstate/RenderStateExtensions.java:
```diff
@@ -0,0 +1,83 @@
+        var modifiers = (Collection<BiConsumer<E, S>>) (Object) ENTITY_CACHE.computeIfAbsent((Class<? extends EntityRenderer<E, S>>) renderer.getClass(), aClass -> {
+            var builder = ImmutableList.<BiConsumer<?, ?>>builder();
+            for (var entry : ENTITY.entrySet()) {
+                if (aClass.isAssignableFrom(entry.getKey())) {
```
```suggestion
                if (entry.getKey().isAssignableFrom(aClass)) {
```
This needs to be mirrored, it currently checks whether the incoming renderer's class is the same as or a superclass of the class the modifier is registered for.

dhyces replied:
Ah, right, yeah. Not sure how I mixed that up...

XFactHD reviewed (changes requested):
While testing, I found that registering modifiers for generic renderers leads to really ugly, unchecked casts in the registration. For example, registering a modifier for any `LivingEntityRenderer` ends up with this:
```java
event.registerEntityModifier(
    (Class<? extends EntityRenderer<LivingEntity, LivingEntityRenderState>>) LivingEntityRenderer.class,
    (entity, renderState) -> {}
);
```
since you can't specify generics in a class literal. I'm not sure there is anything we can realistically do about this though without just entirely dropping type safety.

dhyces committed:
Fix flipped method call

dhyces committed:
Update docs

dhyces committed:
Add living entity register method

Acts as an example for users

dhyces committed:
Some more jdoc changes

dhyces committed:
Update to using TypeTokens

XFactHD commented:
@neoforged/bots run applyAllFormatting

lukebemish commented on src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java:
```diff
@@ -0,0 +1,74 @@
+     * @param <S>          The specific render state type
+     */
+    public <E extends Entity, S extends EntityRenderState> void registerEntityModifier(TypeToken<? extends EntityRenderer<E, S>> baseRenderer, BiConsumer<E, S> modifier) {
+        RenderStateExtensions.registerEntity(baseRenderer.getRawType(), modifier);
```
Isn't this a bit unsafe? After all, if this is passed with a type-token for a class with a _specific_ generic type, this will register the modifier for _all_ renderers of that raw type. That is, like, imagine someone calls this with a `BiConsumer<A, B>` and a type-token for `SomeGenericThingy<A, B>`; this will end up registering that modifier for _all_ `SomeGenericThingy`, including, say, a `SomeGenericThingy<C, D>` -- resulting in a ClassCastException at runtime when the `BiConsumer` is executed!

lukebemish replied:
Just to follow up on this, the example would be if someone did, I dunno, something like
```java
event.registerEntityModifier(
    new TypeToken<LivingEntityRenderer<Mob, LivingEntityRenderState, ?>>(){}
    (entity, renderState) -> {
        Mob mob = entity; // this does not require a cast, the type of `entity` is `Mob`
        // Does something that assumes `entity` is a `Mob`
    }
);
```
This code does not have any sort of unchecked casts from the end of the person writing it, and you might imagine someone could use this if they wanted to modify all renderers registered for `Mob`s! _However_, this would in fact register that modifier for _every_ `LivingEntityRenderer`, resulting in `ClassCastException`s at runtime. This is an enormous footgun -- runtime type errors are... not great. Personally, I'd say it's best to either (a) require a `Class`, so the fundamentally unsafe behavior is locked behind an unsafe cast, or (b) do validation that the `E` and `S` in question -- recovered from the TypeToken, which is its own bit of fun -- are exactly equal to the upper bounds of the class provided, to avoid this sort of issue. The second is still sub-optimal as its a runtime validation, but at least it's a runtime error that will _always_ occur instead of risking only being discovered with the wrong mod combination.

Additionally, I would say it's worth keeping around an overload with `Class` for simplicity's sake in the cases where that is enough.

lukebemish commented on src/main/java/net/neoforged/neoforge/client/renderstate/RenderStateExtensions.java:
```diff
@@ -0,0 +1,83 @@
+    private RenderStateExtensions() {}
+
+    private static final Map<Class<?>, Collection<BiConsumer<?, ?>>> ENTITY = new Reference2ObjectArrayMap<>();
+    private static final Map<Class<?>, Collection<BiConsumer<?, ?>>> ENTITY_CACHE = new Reference2ObjectOpenHashMap<>();
```
Is there a reason this uses `Reference2ObjectArrayMap`? There's a fair number of entity renderer subclasses in vanilla alone (22) -- and this `ENTITY_CACHE` will have an entry for each of them, meaning that the worst-case you are iterating the entire map, and on average you're iterating 11 things, plus whatever mods add. It feels strange that an array map would be faster at this than an identity hash map? Please correct me if I'm wrong.

XFactHD replied:
The `ENTITY` map is only ever iterated outside of registration and the `ENTITY_CACHE` is a `Reference2ObjectOpenHashMap` and is the one that is actually queried when a renderer computes the render state.

lukebemish replied:
Nevermind, I misread this and was looking at the line above

dhyces committed:
Apply formatting

dhyces committed:
Adjust test scaling

dhyces committed:
Remove redundant cast

dhyces committed:
Work on runtime type safety checks

Not at all finished

lukebemish commented on src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java:
```diff
@@ -0,0 +1,112 @@
+                var typeArgs = ((Class) parameterizedType.getRawType()).getTypeParameters();
+                for (int i = 0; i < userArgs.length; i++) {
+                    var userType = userArgs[i];
+                    // ignore wildcards. compilation would catch illegal states, so this must not be an important type to capture
```
Uh, this is false -- compilation would _not_ catch illegal states. If you have a `? extends Mob` that's a wildcard type, will compile fine, will imply it's a Mob, but could be a living entity at runtime if it's a parameter in a LivingEntityRenderer.

lukebemish commented on src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java:
```diff
@@ -0,0 +1,112 @@
+                        var userBounds = typeVariable.getBounds();
+                        // check interface bounds
+                        for (int j = 0; j < typeBounds.length; j++) {
+                            if (!((Class) typeBounds[j]).isAssignableFrom((Class<?>) userBounds[j])) {
```
Why are these Class casts safe? I'm pretty sure nothing guarantees that those won't be some other Type.

lukebemish commented on src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java:
```diff
@@ -0,0 +1,112 @@
+                        }
+                    } else {
+                        // check the base class is the same
+                        if (typeBounds.length != 1 || typeBounds[0] != userType) {
```
I'd use `.equals` here rather than `==` -- remember the bounds could themselves be parameterized types, not necessarily Classes

lukebemish reviewed (changes requested):
I'd make use of type tokens for checking compatibility in that check -- the current logic has some substantial holes in that and a type token super/subclass comparison lets you avoid all those. If not, please make sure there's no wacky edge cases here in any case -- that's a subtle bug caused by some particular mod interaction waiting to happen. If you're not making use of compile time type safety, the runtime type safety needs to be quite strong so it's guaranteed to fire in dev if something could go wrong.

lukebemish commented on src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java:
```diff
@@ -0,0 +1,112 @@
+                    if (userType instanceof WildcardType) continue;
+                    var typeBounds = typeArgs[i].getBounds();
+                    // a generic class with multiple bounds, ie AbstractHoglinRenderer<T extends Mob & HoglinBase>
+                    if (userType instanceof TypeVariable<?> typeVariable) {
```
To clarify the issue with this logic, let's say you do 
```java
static class HoglinRendererTypeToken<T extends Mob & HoglinBase> extends TypeToken<AbstractHoglinRenderer<T>> {}
```
And then call 
```java
event.<Zoglin, HoglinRenderState>registerEntityModifier(new HoglinRendererTypeToken<>(), (entity, renderState) -> {
    Zoglin zoglin = entity; // this is safe
});
```
The current verification code will pass this; it is very much not typesafe.

dhyces committed:
Adjust type safety (big thanks to @lukebemish)

dhyces committed:
Example usage docs

dhyces committed:
Convenience method

Fuzss commented on patches/net/minecraft/client/renderer/entity/EntityRenderer.java.patch:
```diff
@@ -9,6 +9,14 @@
+     public final S createRenderState(T p_361382_, float p_360885_) {
+         S s = this.reusedState;
+         this.extractRenderState(p_361382_, s, p_360885_);
++        net.neoforged.neoforge.client.renderstate.RenderStateExtensions.onUpdateEntityRenderState(this, p_361382_, s);
```
I've worked with this system for a bit in a mod of mine that renders health bars above mobs and therefore must add the relevant information to the render state. 

From that experience I feel like it would be more useful here to instead fire a new all purpose event for modifying the created render state, offering the ability to both modify existing render state values set by vanilla, and for manually adding new values via `IRenderStateExtension`.

This new event could exist in addition to `RegisterRenderStateModifiersEvent`, but could also outright replace it, since I'm not sure it's really necessary here to take the approach of registering a bunch of filters and consumers. Just letting modders handle things when the render state is actually created should work fine.

The new event could also be a potential replacement for `RenderNameTagEvent$CanRender` as it would also allow for overriding the name tag set to the render state.

dhyces replied:
You can still modify vanilla values with this system, its just not guaranteed to stay that same value. In fact, even with an event it wouldn't be guaranteed. Additionally, the idea of firing another event every render frame for every entity was shut down repeatedly for performance concerns, which is why you see this implementation instead.

dhyces committed:
Update test

dhyces committed:
Merge branch '1.21.x' into feat/render-state-event

XFactHD commented on tests/src/main/java/net/neoforged/neoforge/debug/client/ClientEventTests.java:
```diff
@@ -126,4 +140,47 @@ static void renderPlayerEvent(final DynamicTest test) {
+        var numRenderAttachmentKey = new ContextKey<Integer>(ResourceLocation.fromNamespaceAndPath(test.createModId(), "times_to_render"));
+        var testAttachment = test.registrationHelper().attachments().registerSimpleAttachment("test", () -> 3);
+        test.whenEnabled(listeners -> {
+            listeners.mod().addListener((RegisterRenderStateModifiersEvent event) -> {
+                event.registerEntityModifier(PlayerRenderer.class, (entity, renderState) -> {
+                    renderState.setRenderData(rotationKey, 45f);
+                });
+                event.registerEntityModifier(new TypeToken<LivingEntityRenderer<? extends LivingEntity, LivingEntityRenderState, ?>>() {}, (entity, renderState) -> {
+                    renderState.setRenderData(numRenderAttachmentKey, entity.getData(testAttachment));
+                });
+                // Test other type parameters for safety
+                event.registerEntityModifier(new TypeToken<AbstractHoglinRenderer<?>>() {}, (entity, renderState) -> {});
+                event.registerEntityModifier(new TypeToken<MobRenderer<Mob, LivingEntityRenderState, ?>>() {}, (entity, renderState) -> {});
+                try {
+                    class TestBrokenHoglinRendererTypeToken<T extends Mob & HoglinBase> extends TypeToken<AbstractHoglinRenderer<T>> {}
+                    event.registerEntityModifier(new TestBrokenHoglinRendererTypeToken<>(), (entity, renderState) -> {});
+                    test.fail("Unsafe type parameter succeeded. Cannot assume T can be ?.");
+                } catch (IllegalArgumentException ignored) {}
+            });
```
The modifier registration must be moved outside the `test.whenEnabled()` lambda and registered via `test.framework().modEventBus()` instead, otherwise the modifiers won't be present when the test is enabled after joining a world, causing the test to fail.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java:
```diff
@@ -0,0 +1,142 @@
+                    for (var singleBound : typeArg.getBounds()) {
+                        var token = Container.of(TypeToken.of(singleBound));
+                        if (!token.isSubtypeOf(userToken)) {
+                            throw new IllegalArgumentException("%s does not match actual type parameter %s".formatted(userArg, singleBound));
```
```suggestion
                            throw new IllegalArgumentException("%s does not match expected type parameter %s".formatted(userArg, singleBound));
```

dhyces committed:
Apply feedback

dhyces committed:
Merge branch 'feat/render-state-event' of https://github.com/dhyces/NeoForge into feat/render-state-event

XFactHD commented on tests/src/main/java/net/neoforged/neoforge/debug/client/ClientEventTests.java:
```diff
@@ -126,4 +140,47 @@ static void renderPlayerEvent(final DynamicTest test) {
+        var rotationKey = new ContextKey<Float>(ResourceLocation.fromNamespaceAndPath(test.createModId(), "rotation"));
+        var numRenderAttachmentKey = new ContextKey<Integer>(ResourceLocation.fromNamespaceAndPath(test.createModId(), "times_to_render"));
+        var testAttachment = test.registrationHelper().attachments().registerSimpleAttachment("test", () -> 3);
+        test.eventListeners().mod().addListener((RegisterRenderStateModifiersEvent event) -> {
```
```suggestion
        test.framework().modEventBus().addListener((RegisterRenderStateModifiersEvent event) -> {
```
This must use `test.framework().modEventBus()` instead of `test.eventListeners().mod()`, otherwise the listener won't be invoked.

dhyces replied:
Huh. Changed it, a little annoying that there are two similar ways that seem to be correct, but one is only active when the test is enabled...

dhyces committed:
Change which event group is used

Co-authored-by: Dennis C <xfacthd@gmx.de>

dhyces committed:
Merge branch '1.21.x' into feat/render-state-event

XFactHD reviewed (dismissed):
Looks good to me now and works as expected in the tests.
I would appreciate a second set of eyes on this though.

Technici4n commented on src/main/java/net/neoforged/neoforge/client/renderstate/BaseRenderState.java:
```diff
@@ -0,0 +1,43 @@
+ * Allows modders to add arbitrary data onto render states for use in custom rendering.
+ */
+public abstract class BaseRenderState implements IRenderStateExtension {
+    protected Map<ContextKey<?>, Object> extensions = new Reference2ObjectOpenHashMap<>();
```
no final?

Matyrobbrt commented on tests/src/main/java/net/neoforged/neoforge/debug/client/ClientEventTests.java:
```diff
@@ -126,4 +140,47 @@ static void renderPlayerEvent(final DynamicTest test) {
+                }
+                poseStack.popPose();
+            });
+            test.pass();
```
Succeeding immediately on enabled? this sounds wrong

dhyces committed:
Merge remote-tracking branch 'upstream/1.21.x' into feat/render-state-event

dhyces committed:
Fix patch

dhyces committed:
Fix test

dhyces committed:
Finalize extensions field

XFactHD reviewed (approved):
:shipit:
